### PR TITLE
Optionally warn about redundant annotations

### PR DESCRIPTION
**Pull request by AndrewShf**
When visiting each type use location, we get the default annotation of the type or variable by using `addComputedTypeAnnotations(Tree, AnnotatedTypeMirror, Boolean useflow=false)` and compare it with the annotation got from `getExplicitAnnotation()`

### CODE REVIEW

**Comment by wmdietl** [comment]
Thanks!
I think we should make this check optional - to avoid the test failures, to avoid the performance hit, and to not start breaking users without warning.

The most interesting failure I saw was the crash here: https://dev.azure.com/eisop/checker-framework/_build/results?buildId=1758&view=logs&j=d9c91959-2a76-59b4-c041-71226f63c7f7&t=8e32d791-4920-5582-276c-9e9e2fd189a8&l=555

**Comment by AndrewShf** [comment]
> Thanks! I think we should make this check optional - to avoid the test failures, to avoid the performance hit, and to not start breaking users without warning.
> 
> The most interesting failure I saw was the crash here: https://dev.azure.com/eisop/checker-framework/_build/results?buildId=1758&view=logs&j=d9c91959-2a76-59b4-c041-71226f63c7f7&t=8e32d791-4920-5582-276c-9e9e2fd189a8&l=555

This is because currently we don't annotate default on type variables, as says in 
https://github.com/eisop/checker-framework/blob/master/framework/src/main/java/org/checkerframework/framework/util/defaults/QualifierDefaults.java#L930-L935
This method is called by `QualifierDefaults.java#applyDefaultsElement` https://github.com/eisop/checker-framework/blob/d17ecdf64414962f5b4686a53dcf9bf2547c1d30/framework/src/main/java/org/checkerframework/framework/util/defaults/QualifierDefaults.java#L845

**Comment by zcai1** [comment]
LGTM

File: docs/CHANGELOG.md [hunk 1]
```diff
@@ -3,6 +3,10 @@ Version 3.26.0-eisop2 (October ?, 2022)
 
 **User-visible changes:**
 
+The new command-line argument `AcheckRedundantAnnotations` helps to reduce redundant annotations.
+When supplying this flag, program will issue a warning if the explicitly written annotation
+on a type is the same as the default annotation would be applied on this type.
```
Comment (wmdietl) [inline docs/CHANGELOG.md:8]:
```suggestion
With this flag, a warning is issued if an explicitly written annotation on a type is the same as the default
annotation for this type and location.
```

File: checker/tests/nullness-checkredundantannotations/RedundantAnnotation.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        // expects a "redundant.anno" warning on the typecast
+        String x = (@Nullable String) obj;
+
+        // expects a "redundant.anno" warning on the typecast
+        // :: error: (instanceof.nullable)
+        boolean b = x instanceof @Nullable String;
+
+        // expects a "redundant.anno" warning on the component type
+        @Nullable String[] strs = new String[10];
+    }
```
Comment (wmdietl) [inline checker/tests/nullness-checkredundantannotations/RedundantAnnotation.java:82]: Can you add a test case where you e.g. use `@DefaultQualifier(Nullable.class)` in a scope and use a redundant '@Nullable` on e.g. a parameter?

File: checker/tests/nullness-checkredundantannotations/RedundantAnnotation.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        Object obj = null;
+        // expects a "redundant.anno" warning on the typecast
+        String x = (@Nullable String) obj;
+
+        // expects a "redundant.anno" warning on the typecast
+        // :: error: (instanceof.nullable)
+        boolean b = x instanceof @Nullable String;
+
+        // expects a "redundant.anno" warning on the component type
+        @Nullable String[] strs = new String[10];
```
Comment (wmdietl) [inline checker/tests/nullness-checkredundantannotations/RedundantAnnotation.java:81]: Is the default component type of an array `@Nullable`? I would have expected `@NonNull`, at least on the LHS. Maybe on the RHS, in the array creation, the default is `@Nullable`?

File: checker/tests/nullness-checkredundantannotations/RedundantAnnotation.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        try (@Nullable InputStream in = arg) {
+
+            // :: warning: (redundant.anno)
+            // :: warning: (nullness.on.exception.parameter)
+        } catch (@NonNull Exception e) {
+
+        }
+
+        // expects a "redundant.anno" warning on the upper bound
+        List<? extends @Nullable Object> l;
```
Comment (wmdietl) [inline checker/tests/nullness-checkredundantannotations/RedundantAnnotation.java:68]: This seems wrong: There is a difference between `List<? extends @Nullable Object>` and `List<? extends Object>'. Did you mean `List<? extends @NonNull Object>`?

File: checker/tests/nullness-checkredundantannotations/RedundantAnnotation.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        // expects a "redundant.anno" warning on the lower bound
+        List<? super @Nullable Object> l2;
```
Comment (wmdietl) [inline checker/tests/nullness-checkredundantannotations/RedundantAnnotation.java:70]: I would have expected `@NonNull` as the default lower bound.

File: docs/manual/introduction.tex [hunk 6]
```diff
@@ -721,6 +721,10 @@
   JDK, but could also be useful when the warning/note is redundant for a
   user.
   See comments for \<-ApermitUnsupportedJdkVersion> option above.
+\item \<-AcheckRedundantAnnotations>
```
Comment (wmdietl) [inline docs/manual/introduction.tex:724]: This seems like the best section for this new option, but I would move it earlier in the list, probably as the second one, between `-Alint` and `-AsuggestPureMethods`.

File: docs/manual/introduction.tex [hunk 7]
```diff
@@ -721,6 +721,10 @@
...
+  Supplying this flag helps to reduce redundant annotations,
```
Comment (wmdietl) [inline docs/manual/introduction.tex:725]: Same change as in changelog.

File: docs/manual/introduction.tex (hunk 6 above)
Comment (wmdietl) [inline docs/manual/introduction.tex:724]: Did we already brainstorm option names?
What do you think of `warnRedundantAnnotations` (similar to `-AwarnUnneededSuppressions`)?
The other `-AcheckXXX` options are more about additional checks... but then again, what is the difference between additional checks and additional warnings? So what sounds better?

File: framework/src/main/java/org/checkerframework/common/basetype/BaseTypeValidator.java [hunk 8]
```diff
@@ -674,7 +674,6 @@ public Void visitWildcard(AnnotatedWildcardType type, Tree tree) {
         if (!areBoundsValid(type.getExtendsBound(), type.getSuperBound())) {
             reportInvalidBounds(type, tree);
         }
-
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/common/basetype/BaseTypeValidator.java:677]: I would probably undo this unrelated change.

**Review by wmdietl — COMMENTED** [review]
Some first comments, will look at `BaseTypeVisitor.java` later.

File: checker/tests/nullness-checkredundantannotations/RedundantAnnotation.java (hunk 3 above)
Comment (AndrewShf) [reply checker/tests/nullness-checkredundantannotations/RedundantAnnotation.java:82]: Both should be `@NonNull`. I have corrected it.

File: docs/manual/introduction.tex (hunk 6 above)
Comment (AndrewShf) [reply docs/manual/introduction.tex:724]: I think `warnRedundantAnnotations` is more straightforward and easier to understand. `checkXX` doesn't say what we check for.

File: checker/tests/nullness-checkredundantannotations/RedundantAnnotation.java (hunk 4 above)
Comment (AndrewShf) [reply checker/tests/nullness-checkredundantannotations/RedundantAnnotation.java:69]: Yes, I have corrected it.

File: checker/tests/nullness-checkredundantannotations/RedundantAnnotation.java (hunk 5 above)
Comment (AndrewShf) [reply checker/tests/nullness-checkredundantannotations/RedundantAnnotation.java:71]: yes. I have fixed this.

File: checker/tests/nullness-checkredundantannotations/RedundantAnnotation.java (hunk 2 above)
Comment (AndrewShf) [reply checker/tests/nullness-checkredundantannotations/RedundantAnnotation.java]: added

File: framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java [hunk 9]
```diff
@@ -1489,9 +1494,39 @@ && getCurrentPath().getParentPath().getLeaf().getKind()
...
 
+    /**
+     * Issues a "redundant.anno" warning if the annotation written on the type is the same as the
+     * default annotation for this type and location.
+     *
+     * @param tree an AST node
+     * @param type get the explicit annotation on this type and compare it with the default one for
+     *     this type and location.
+     */
+    void warnRedundantAnnotations(Tree tree, AnnotatedTypeMirror type) {
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java:1509]:
```suggestion
    protected void warnRedundantAnnotations(Tree tree, AnnotatedTypeMirror type) {
```

**Review by wmdietl — APPROVED** [review]
Thanks for this new feature!
